File: chapter2_cpp/dsl/Order.cpp
#include "Order.h"
#include "OrderBuilder.h"

namespace dsl {
    Order::Order(const OrderBuilder& b)
        : _security(b.security()),
        _quantity(b.quantity()),
        _limitPrice(b.limitPrice()),
        _allOrNone(b.allOrNone()),
        _value(b.value()),
        _boughtOrSold(b.boughtOrSold())
    {
    }
    std::ostream& operator<<(std::ostream& os, const Order& order)
    {
        os << "security : " << order.security() << std::endl
            << "quantity : " << order.quantity() << std::endl
            << "limitPrice : " << order.limitPrice() << std::endl
            << "allOrNone : " << order.allOrNone() << std::endl
            << "value : " << order.value() << std::endl;
        return os;
    }
}

File: chapter2/Order.fs
module Order 
type BuyOrSellType =
    | Buy
    | Sell
    override x.ToString() =
        match x with
        | Buy -> "Buy"
        | Sell -> "Sell"

type Order =
    { Security : string;
        Quantity : int;
        LimitPrice : int;
        BuyOrSell : BuyOrSellType;
    }
    static member Default = 
        { Security = "";
          Quantity = 0;
          LimitPrice = 0;
          BuyOrSell = Buy;
        }


    member x.buy(security, quantity) =
        { Security = security;
          Quantity = quantity;
          LimitPrice = x.LimitPrice;
          BuyOrSell = Buy;
        }

    member x.sell(security, quantity) =
        { Security = security;
          Quantity = quantity;
          LimitPrice = x.LimitPrice;
          BuyOrSell = Sell;
        }

    member x.limitPrice(price) =
        { Security = x.Security;
          Quantity = x.Quantity;
          LimitPrice = price;
          BuyOrSell = Sell;
        }

    member x.valueAs(f) : int =
        f x.Quantity x.LimitPrice

